In [ ]:
#Notas Semana 6

**Actividad.** Mónica hace aretes y cadenitas de joyería. Es tan buena, que todo lo que hace lo vende.

Le toma 30 minutos hacer un par de aretes y una hora hacer una cadenita, y como Mónica también es estudiante, solo dispone de 10 horas a la semana para hacer las joyas. Por otra parte, el material que compra solo le alcanza para hacer 15 unidades (el par de aretes cuenta como unidad) de joyas por semana.

La utilidad que le deja la venta de las joyas es $\$15$ en cada par de aretes y $\$20$ en cada cadenita.

¿Cuántos pares de aretes y cuántas cadenitas debería hacer Mónica para maximizar su utilidad?

Formular el problema en la forma explicada y obtener la solución gráfica (puede ser a mano).

**Diez minutos: quien primero lo haga, pase al frente para explicarlo y le subiré la nota de alguna tarea a 100.**

**Solución:**

Se desea maximizar la ganancia. Se sabe que por cada par de aretes, se obtienen $\$15$ y por cada cadena, $\$20$; se define entonces:
* $x_a$: no. de par de aretes
* $x_c$: no. de cadenitas

**Criterio de Optimización:**
$$
\max [15\quad20]\begin{bmatrix}x_a\\x_c\end{bmatrix}=\min
[-15\quad-20]\begin{bmatrix}x_a\\x_c\end{bmatrix}
$$

``` #'bmatrix' para que sea una matriz de corchetes; 'pmatrix' para que sea con paréntesis ```

sujeto a las siguientes restricciones:
* De tiempo: $0.5x_a+x_c\leq10$
* De material: $x_a+x_c\leq15$.

In [ ]:
import numpy as np
import scipy.optimize as opt

In [ ]:
c=np.array([-15,-20])
A=np.array([[0.5,1],[1,1]])
b=np.array([10,15])
opt.linprog(c,A,b)

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: -250.0
              x: [ 1.000e+01  5.000e+00]
            nit: 2
          lower:  residual: [ 1.000e+01  5.000e+00]
                 marginals: [ 0.000e+00  0.000e+00]
          upper:  residual: [       inf        inf]
                 marginals: [ 0.000e+00  0.000e+00]
          eqlin:  residual: []
                 marginals: []
        ineqlin:  residual: [ 0.000e+00  0.000e+00]
                 marginals: [-1.000e+01 -1.000e+01]
 mip_node_count: 0
 mip_dual_bound: 0.0
        mip_gap: 0.0

**Respuesta:**

Se necesitan hacer 10 pares de aretes y 5 cadenitas para obtener una ganancia de $\$250$.
___

## 4. Problema de transporte 1
- **Referencia**: http://bdigital.unal.edu.co/5037/4/guillermojimenezlozano.2006_Parte1.pdf

Una  empresa tiene dos factorías A y B.  En ellas  se fabrica un determinado  producto, a razón de 500  y 400 unidades por día respectivamente.  El producto  ha  de  ser distribuido posteriormente  a  tres  centros  C,  D  y  E,  que  requieren,  respectivamente,  200,  300  y  400  unidades.  Los  costos  de  transportar cada unidad del producto desde cada factoría a cada centro distribuidor son los  indicados  en la tabla  siguiente:

Factoría|C|D|E|Fabricación (Unidades)
:----|----|----|----|----
A|  50 u.m|60 u.m|10 u.m|600 u
B|  25 u.m|40 u.m|20 u.m|500 u
Demanda|200|300|400|

**¿De qué manera  deben  organizar  el transporte a fin de que los gastos sean mínimos?**


Formulemos el problema para ser resuelto por la programación lineal con
- $x_1$: unidades transportadas de la factoría "A" al centro "C"
- $x_2$: unidades transportadas de la factoría "A" al centro "D"
- $x_3$: unidades transportadas de la factoría "A" al centro "E"
- $x_4$: unidades transportadas de la factoría "B" al centro "C"
- $x_5$: unidades transportadas de la factoría "B" al centro "D"
- $x_6$: unidades transportadas de la factoría "B" al centro "E"

**Solución:**

Se desea minimizar gastos de transporte, donde los costos se presentan por unidad de cada fábrica a cada centro (lo que corresponde a la definición de las variables), por lo que el criterio de optimización es:
$$
\min[50,60,10,\;25,40,20]x,
$$
donde $x=[x_1,x_2,x_3,x_4,x_5,x_6]^T$. Bajo las siguientes restricciones:
* De fabricación:
  - fábrica **A**: $x_1+x_2+x_3\leq600$,
  - fábrica **B**: $x_4+x_5+x_6\leq500$
* De demanda:
  - centro **C**: $x_1+x_4\geq200$,
  - centro **D**: $x_2+x_5\geq300$,
  - centro **E**: $x_3+x_6\geq400$.


In [ ]:
c=np.array([50,60,10,25,40,20])
A=np.array([[1,1,1,0,0,0],
           [0,0,0,1,1,1],
           [-1,0,0,-1,0,0],
           [0,-1,0,0,-1,0],
           [0,0,-1,0,0,-1]])
b=np.array([600,500,-200,-300,-400])

opt.linprog(c,A,b)

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: 21000.0
              x: [ 0.000e+00  0.000e+00  4.000e+02  2.000e+02  3.000e+02
                   0.000e+00]
            nit: 3
          lower:  residual: [ 0.000e+00  0.000e+00  4.000e+02  2.000e+02
                              3.000e+02  0.000e+00]
                 marginals: [ 2.500e+01  2.000e+01  0.000e+00  0.000e+00
                              0.000e+00  1.000e+01]
          upper:  residual: [       inf        inf        inf        inf
                                    inf        inf]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              0.000e+00  0.000e+00]
          eqlin:  residual: []
                 marginals: []
        ineqlin:  residual: [ 2.000e+02  0.000e+00  0.000e+00  0.000e+00
                              0.000e+00]
                 marginals: [-0.000e+00 -0.000e+00 -2.5

In [ ]:
# Fábrica A manda 0 a C, 0 a D, 400 a E
# Fábrica B manda 200 a C, 300 a D, 0 a E

**Respuesta:**

El gasto mínimo es de 21000 u.m., con la siguiente distribución:
* Fábrica **A**: 400 unidades al centro **E**,
* Fábrica **B**: 200 unidades al centro **C** y 300 al centro **D**.
___

## 5. Otro problema de dieta
A una persona que quiere adelgazar se le ofrecen dos productos, $A$ y $B$, para que tome una mezcla de ambos con las siguientes recomendaciones:
* No debe tomar más de $150$ g de la mezcla ni menos de $50$ g.
* La cantidad de $A$ debe ser igual o superior a la de $B$.
* No debe incluir más de $100$ g de $A$.

Si $100$ g de $A$ contiene $30$ mg de vitaminas y $450$ calorías y, $100$ g de $B$ contienen 20 mg de vitaminas y $150$ calorías:
* ¿Cuántos gramos de cada producto debe mezclar para obtener el preparado más rico en
vitaminas?
* ¿Y el más pobre en calorías?

**Primer objetivo: maximizar las vitaminas**

Se sabe que la composición de **A** es de $30\;\mathrm{mg}$ de vitaminas por cada $100\;\mathrm{g}$ de éste, y la composición de **B** es de $20\;\mathrm{mg}$ por cada $100\;\mathrm{g}$. Se definen entonces las variables siguientes:
* $x_A$ como la cantidad en gramos de vitaminas en cada $100\;\mathrm{g}$ del compuesto **A** ,
* $x_B$ como la cantidad en gramos de vitaminas en cada $100\;\mathrm{g}$ del compuesto **B**.

Por lo que el criterio de optimización es el siguiente:
$$costo =\min_{x_A,x_B}(-0.030x_A-0.020x_B),$$
sujeto a las siguientes restricciones:
1. $x_A+x_B\leq{150}$
2. $x_A+x_B\geq{50}$
3. $x_A\geq{x_B}$
4. $x_A\leq{100}$

y para poder resolver con linprog, debo acomodarlas todas de modo que queden con un $\leq{}$, por lo que replanteo ecuaciones 2 y 3:

2. $-x_A-x_B\leq{-50}$
3. $-x_A+x_B\leq{0}$

equivaliendo 1 y 2 al rango de la mezcla, 3 a la relación entre A y B, y 4 al límite del compuesto A.

In [1]:
import numpy as np
import scipy.optimize as opt

In [4]:
c=np.array([-0.030,-0.020])
A=np.array([[1,1],[-1,-1],[-1,1],[1,0]])
b=np.array([150,-50,0,100])
opt.linprog(c,A,b)

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: -4.0
              x: [ 1.000e+02  5.000e+01]
            nit: 2
          lower:  residual: [ 1.000e+02  5.000e+01]
                 marginals: [ 0.000e+00  0.000e+00]
          upper:  residual: [       inf        inf]
                 marginals: [ 0.000e+00  0.000e+00]
          eqlin:  residual: []
                 marginals: []
        ineqlin:  residual: [ 0.000e+00  1.000e+02  5.000e+01  0.000e+00]
                 marginals: [-2.000e-02 -0.000e+00 -0.000e+00 -1.000e-02]
 mip_node_count: 0
 mip_dual_bound: 0.0
        mip_gap: 0.0

```# El fun: -4.0 me dice que estoy consumiendo en total 4 g de vitaminas```
```# El x: [100 50] me dice que la mezcla que maximiza el consumo de vitaminas es de 100 g del A y 50 de B```
```# El  ineqlin:  residual: [ 0.000e+00  1.000e+02  5.000e+01  0.000e+00] me dice que```
```# (0) estoy en el máximo (consumiendo 150 g de mezcla)```
```# (100) estoy 100 g por encima del mínimo (xa+xb >= 50)```
```# (50) A es 50 g mayor q B y (0) que xa=100 g exactamente.```

**Resultado**

La mezcla que maximiza el consumo de vitaminas es de $100\;\mathrm{g}$ del compuesto **A** y $50\;\mathrm{g}$ del compuesto **B**, consumiendo $4\;\mathrm{g}$ de vitaminas.

Adicionalmente, se observa que, de la mezcla de $100\;\mathrm{g}$ más del rango inferior y $50\;\mathrm{g}$ más en la relación entre **A** y **B**.
___

**Segundo Objetivo:**

Conservando las variables establecidas $x_A$ y $x_B$, para minimizar la cantidad de calorías de la mezcla, el criterio de optimización resulta:
$$\min_{x_A,x_B}(4.5x_A+1.5x_B),$$
bajo las mismas restricciones

In [5]:
c=np.array([4.5,1.5])
opt.linprog(c,A,b)

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: 150.0
              x: [ 2.500e+01  2.500e+01]
            nit: 2
          lower:  residual: [ 2.500e+01  2.500e+01]
                 marginals: [ 0.000e+00  0.000e+00]
          upper:  residual: [       inf        inf]
                 marginals: [ 0.000e+00  0.000e+00]
          eqlin:  residual: []
                 marginals: []
        ineqlin:  residual: [ 1.000e+02  0.000e+00  0.000e+00  7.500e+01]
                 marginals: [-0.000e+00 -3.000e+00 -1.500e+00 -0.000e+00]
 mip_node_count: 0
 mip_dual_bound: 0.0
        mip_gap: 0.0

**Resultado:**

El consumo mínimo de calorías es de 150 con 25 g de A y 25 de B. El rango de la mezcla está 100 g por debajo del límite superior (```1.000e+02```); la diferencia entre A y B es cero y el compuesto A está 75 g por debajo de su límite máximo (```7.500e+01```).
___